weather/weather.go

```diff
@@ -12,56 +10,51 @@ const (
 	weatherURL  = "https://api.openweathermap.org/data/2.5/weather?appid="
 )
 
-type Client struct {
+type WeatherClient struct {
 	apiKey string
 	unit   string
 
-	common service
-
-	Temperature *TemperatureService
-	Location    *LocationService
+	httpClient *resty.Client
 }
 
-type service struct {
-	client *Client
+type WeatherConfig struct {
+	ApiKey string
+	Units  string "metric" // This isn't setting a default
 }
 
-func NewClient(apiKey string, unit string) *Client {
-	c := &Client{
-		apiKey: apiKey,
-		unit:   unit,
-	}
-	if c.unit == "" {
-		c.unit = "metric"
+func NewClient(config WeatherConfig) (*WeatherClient, error) {
+
+	if config.ApiKey == "" {
+		cause := errors.New(apiKeyNotSupplied)
+		return nil, errors.WithStack(cause)
 	}
-	c.initialize()
-	return c
-}
 
-func (c *Client) initialize() {
-	c.common.client = c
+	if config.Units == "" {
+		config.Units = "metric"
+	}
 
-	c.Temperature = (*TemperatureService)(&c.common)
-	c.Location = (*LocationService)(&c.common)
+	c := &WeatherClient{
+		apiKey: config.ApiKey,
+		unit:   config.Units,
+	}
+	c.httpClient = resty.New()
+	return c, nil
 }
 
-// TODO Switch to using resty
-func (c *Client) NewRequest(url string) ([]byte, error) {
+func (c *WeatherClient) Fetch(url string) ([]byte, error) {
+
+	resp, err := c.httpClient.R().
+		EnableTrace().
```

Comment:
The `SetResult` function is very useful for resty. It prevents having to unmarshal to JSON as resty can do it for you. Here's an example. https://github.com/SalesLoft/salesloft-go-clients/blob/master/salesloft/api/team.go#L77

Reply:
I spent quite some time on this one, and couldn't figure it out.
`SetResult()` seems to take a pointer as an input.
Since I have a `Fetch()` function handling all my calls, I can't figure out how to expand `Fetch()` to take in any type of struct.

Obviously I could create multiple `Fetch()` functions with different input types, but I'm trying to see if I can wrap my head around interfaces/generics here.